### PR TITLE
build(deps): bump node to v20

### PR DESCRIPTION
**Pull request by yeikel**
Release notes : 

https://nodejs.org/en/blog/announcements/v19-release-announce
https://nodejs.org/en/blog/announcements/v20-release-announce

### CODE REVIEW

**Comment by yeikel** [comment]
@deivid-rodriguez What do you think about this change? 

We're still pinning the npm and yarn versions so it should be fine?

File: npm_and_yarn/Dockerfile [hunk 1]
```diff
@@ -7,7 +7,7 @@ ARG PNPM_VERSION=8.9.2
 ARG YARN_VERSION=3.6.4
 
 # See https://github.com/nodesource/distributions#installation-instructions
-ARG NODEJS_VERSION=18.x
+ARG NODEJS_VERSION=20.x
```
Comment (JamieMagee) [inline npm_and_yarn/Dockerfile:10]: unrelated, but what do you think about moving the `.x` from `NODEJS_VERSION` to line 20 so it's `node_${NODEJS_VERSION}.x`

Comment (yeikel) [reply npm_and_yarn/Dockerfile]: That sounds reasonable, thanks!

**Comment by yeikel** [comment]
> Currently we run a fixed version on nodejs, so that means we are not able to support applications with certain nodejs constraints. For example, if someone specifying "node": "^18.12.0" in their engines package.json section, my understanding is that this PR will break that

As far as I was able to test, it does not seem that NPM enforces this by default

Dependencies can also define what engines they support that can lead to the following warning at installation time  : 

> EBADENGINE Unsupported engine {
> npm WARN EBADENGINE   package: 'pretty-format@28.1.3',
> npm WARN EBADENGINE   required: { node: '^12.13.0 || ^14.15.0 || ^16.10.0 || >=17.0.0' },
> npm WARN EBADENGINE   current: { node: 'v16.1.0', npm: '8.19.2' }
> npm WARN EBADENGINE }

By default this is not enabled but it can be can be enforced via `.npmrc` setting `engine-strict=true which will lead to the following failure at installation time : 

```
npm ERR! code EBADENGINE
npm ERR! engine Unsupported engine
npm ERR! engine Not compatible with your version of node/npm: @jest/console@28.1.3
npm ERR! notsup Not compatible with your version of node/npm: @jest/console@28.1.3
npm ERR! notsup Required: {"node":"^12.13.0 || ^14.15.0 || ^16.10.0 || >=17.0.0"}
npm ERR! notsup Actual:   {"npm":"8.19.2","node":"v16.1.0"}
```

As far as I can tell, we are catching that already when the user enforces it : https://github.com/dependabot/dependabot-core/blob/db9cb4b21a34576e1bc9588ead3b2801018679d2/npm_and_yarn/lib/dependabot/npm_and_yarn/file_updater/npm_lockfile_updater.rb#L422-L426 


> ideally we would be able to use the most appropriate nodejs version for each repo, or even better, package manager would provide a way to ask "please resolve for nodejs version X, regardless of the running node"

I am not sure what you're thinking here, but perhaps we could include more than one version in the image with `nvm` and switch accordingly?

**Review by deivid-rodriguez — APPROVED** [review]

**Comment by yeikel** [comment]
@deivid-rodriguez What changed that you decided to approve it? I am just curious as your points were valid and interesting 

Thank you!

**Comment by deivid-rodriguez** [comment]
Going a bit back and forth with this, but the intention is to deploy this soon-ish, because the potential issues seem quite minor, and not deploying this implies having similar issues for users that require the latest NodeJS version.

I'm still working on going through some common related errors unknown errors and raising them as informative user errors, and I'll probably deploy this once that's done.

**Comment by yeikel** [comment]
@abdulapopoola @jakecoffman Could you please help me to assign a point of contact now that David lost the necessary access to check the metrics (or that's what I understand) ?

**Comment by yeikel** [comment]
@deivid-rodriguez Do you remember where you left off? Thanks!

**Comment by deivid-rodriguez** [comment]
Hei!

I run out of time for working on better error handling for NPM and had to switch to other tasks. So my idea was to ship this as is.

Given your explanation in https://github.com/dependabot/dependabot-core/pull/8275#issuecomment-1785369140 that this is by default a warning and not an error, I think it's fine to move forward with the upgrade. We should not be penalizing users of modern nodejs versions in favor of users of old nodejs versions.

**Comment by yeikel** [comment]
> Hei!
> 
> I run out of time for working on better error handling for NPM and had to switch to other tasks. So my idea was to ship this as is.
> 
> Given your explanation in [#8275 (comment)](https://github.com/dependabot/dependabot-core/pull/8275#issuecomment-1785369140) that this is by default a warning and not an error, I think it's fine to move forward with the upgrade. We should not be penalizing users of modern nodejs versions in favor of users of old nodejs versions.

Awesome, thank you for taking the time to explain. 

Do you recall what was missing in terms of error handling? If you have time at some point in the near future, I'd be nice to log them as tech deb issues 🙏.

 In any case, as you pointed out, this change should be ready to be merged as is. Thank you for taking the time to reply so quickly! 🙇‍♂️ 

@abdulapopoola Any chance you can help move this one forward?

**Comment by deivid-rodriguez** [comment]
I didn't have time to materialize it into a proper issue, but I suspected from analyzing error messages logged in our internal tracker that sometimes we were not properly handling EBADENGINE errors. Because otherwise they should be showing up as user errors and not as unknown errors in our internal tracker.

I think it's fine to wait until someone reports any issues.

**Comment by yeikel** [comment]
@jurre This one is ready as well

**Comment by basil** [comment]
Possibly related to #8896?

**Comment by xu3u4** [comment]
Any plan to merge this PR? 🙏

**Comment by yeikel** [comment]
@jurre Can you please review/help to move this one forward?

**Comment by smoookeeey** [comment]
> @jurre Can you please review/help to move this one forward? 
- 
Approved

**Review by jurre — APPROVED** [review]

**Comment by jurre** [comment]
Thanks @yeikel! I've let this sit in prod for an hour and not observing any issues so far but we'll keep an eye out

**Comment by yeikel** [comment]
> Thanks @yeikel! I've let this sit in prod for an hour and not observing any issues so far but we'll keep an eye out

Thank you! I think that the issues should be minimal because we're unfortunately still pinning to NPM 9.x

**Comment by xu3u4** [comment]
Thank you all 🙇‍♀️🙇‍♀️🙇‍♀️

**Pull request merged**